### PR TITLE
Testcase for #695

### PR DESCRIPTION
**Pull request by dhuebner**
Test case for  Inconsistent handling of synthetic fragment for missing or orphaned obj #695 

The test adds a fake constructor to a JvmGenericType and adds associations.
After removing this constructor from the original resource the resource is stored again causing a "none" fragment created.

Loading the stored bad state causes `null` keys in `org.eclipse.xtext.xbase.jvmmodel.JvmModelAssociator.Adapter.targetToSourceMap`

Storing this state than cause an NPE in `org.eclipse.xtext.xbase.resource.BatchLinkableResourceStorageWritable.logIfResourceMismatch(BatchLinkableResource, EObject)`

### CODE REVIEW

File: org.eclipse.xtext.xbase.tests/src/org/eclipse/xtext/xbase/tests/resources/ResourceStorageTest.java [hunk 1]
```diff
@@ -163,4 +170,47 @@ public void testWriteAndLoad() throws Exception {
...
+		type.getMembers().add(jvmConstructor);
+		associator.associate(file.eResource().getContents().get(0), jvmConstructor);
+
+		StorageAwareResource originalResource = ((StorageAwareResource) file.eResource());
+		ByteArrayOutputStream bout = new ByteArrayOutputStream();
+		resourceStorageFacade.setStoreNodeModel(false);
+		resourceStorageFacade.createResourceStorageWritable(bout).writeResource(originalResource);
+		
+		// Remove synthetic constructor to cause an addition of a "none" fragment
+		type.getMembers().remove(jvmConstructor);
```
Comment (szarnekow) [inline org.eclipse.xtext.xbase.tests/src/org/eclipse/xtext/xbase/tests/resources/ResourceStorageTest.java:195]: @dhuebner I can understand that it'S possible to break the serialization if the model is corrupted. I'd be curious though how that state was "achieved" in the real case.

Comment (dhuebner) [reply org.eclipse.xtext.xbase.tests/src/org/eclipse/xtext/xbase/tests/resources/ResourceStorageTest.java:195]: @szarnekow It happens often when switching a branch, sometimes it also happens during regular work/typing. What I see is that the broken Xtend file has an active annotation or an inner Xtend class has one. `@Data` or `@FinalFieldConstructor` for example manipulating constuctors of the JvmType, so I suppose it's a kind of race condition. The problem is that I can't see why the constructor eResource not in place when a `none` fragment was added.

**Comment by dhuebner** [comment]
@szarnekow 
I prepared a fix for the "none" fragment case.
What would you think about having a logging for the bogus case and trigger further investigation in Xtend with an Active annotation context?

**Comment by szarnekow** [comment]
Sounds good to me. The patch looks fine, too at first glance.

**Comment by dhuebner** [comment]
@szarnekow 
I see a lot of errors logged in my workspace. As we don't know the cause yet and to not annoy the users, I changed log level to `debug`. I also created a new issue in xtext-xtend eclipse/xtext-xtend#1189

@cdietrich @szarnekow 
Could some of you approve this PR?

**Comment by cdietrich** [comment]
TODO create/update downstream builds in eclipse and xtend to see what breaks

**Comment by cdietrich** [comment]
pushed branches

**Comment by dhuebner** [comment]
> TODO create/update downstream builds in eclipse and xtend to see what breaks

Is that a TODO for me?

**Comment by cdietrich** [comment]
> Is that a TODO for me?
i have already done it. (in theory would be a todo for you)

**Comment by dhuebner** [comment]
@cdietrich
>  i have already done it.

Thanks

**Comment by cdietrich** [comment]
no test fails.

**Comment by dhuebner** [comment]
@cdietrich 
I pushed a commit with error logging. I also noticed and fixed a potential NPE. 
... and I don't think it's a useful information that an eProxy doesn't have an eResource

**Comment by cdietrich** [comment]
ok will restart the build once yours went through

**Comment by dhuebner** [comment]
With a change above I see only few error logs. And that all synthetic `JvmConstructor` from Xtend classes that are annotated with `@Data` or `@FilnalFieldConstructor` and are inside an Xtend file but are not the primary Xtend types. Means this types share the same Xtend file with other Xtend types

**Comment by dhuebner** [comment]
@cdietrich 
I see any log entires in jenkins console log

**Comment by cdietrich** [comment]
me neither

**Comment by dhuebner** [comment]
@cdietrich 
Can we merge it?

**Comment by cdietrich** [comment]
i want to wait for @szarnekow am not sure if the error logging is working.

**Comment by dhuebner** [comment]
In xbase the logging is working:
https://ci.eclipse.org/xtext/job/xtext-extras/job/dh-storage-writable-695/6/testReport/org.eclipse.xtext.xbase.tests.resources/ResourceStorageTest/testWriteAndLoadWithMissing/

**Review by szarnekow — APPROVED** [review]

**Comment by szarnekow** [comment]
Looks about right to me.

**Pull request merged**